### PR TITLE
[sw/lib] Delete hmac.h

### PR DESCRIPTION
**Pull request by mcy**
This PR deletes the old hmac.h proto-DIF, and replaces it with calls to the official DIF directly.

Also, this seems to render `sw/vendor/cryptoc` completely unused. Should we delete it?

### CODE REVIEW

File: sw/device/tests/sha256_test.c [hunk 1]
```diff
@@ -21,16 +27,54 @@ static const uint32_t kExpectedDigest[8] = {0xdc96c23d, 0xaf36e268, 0xcb68ff71,
...
+      case kDifHmacDigestProcessing:
+        continue;
+      case kDifHmacDigestOk:
+        goto read_end;
+      default:
+        LOG_ERROR("Error while reading the digest.");
+        return false;
+    }
+  }
+read_end:
```
Comment (gkelly) [inline sw/device/tests/sha256_test.c:69]: This control flow is pretty deep, what about:

```
dif_hmac_result_t result = kDifHmacDigestProcessing;
while (result == kDifHmacDigestProcessing) {
  result = dif_hmac_digest(&hmac, &digest);
}
if (result != kDifHmacDigestOk) {
  LOG_ERROR("Error while reading the digest.");
  return false;
}
```

File: sw/device/tests/sha256_test.c [hunk 2]
```diff
@@ -21,16 +27,54 @@ static const uint32_t kExpectedDigest[8] = {0xdc96c23d, 0xaf36e268, 0xcb68ff71,
...
-      LOG_ERROR("Digest mismatched at index %d: exp: %x, act: %x", i, digest[i],
-                kExpectedDigest[i]);
-      flash_write_scratch_reg(digest[i]);
+  for (uint32_t i = 0; i < 8; ++i) {
+    uint32_t got = digest.digest[i];
+    uint32_t want = kExpectedDigest[i];
+    if (got != want) {
+      LOG_ERROR("Digest mismatch at index %d: wanted 0x%x, got 0x%x.", i, want,
+                got);
+      flash_write_scratch_reg(got);  // ?
```
Comment (gkelly) [inline sw/device/tests/sha256_test.c:77]: The question mark here isn't a useful comment. What about adding a TODO for documenting this?

File: sw/device/tests/sha256_test.c [hunk 3]
```diff
@@ -21,16 +27,54 @@ static const uint32_t kExpectedDigest[8] = {0xdc96c23d, 0xaf36e268, 0xcb68ff71,
...
+        goto push_end;
+      case kDifHmacFifoFull:
+        data += bytes_sent;
+        data_left -= bytes_sent;
+      default:
+        LOG_ERROR("Error while pushing to FIFO.");
+        return false;
+    }
+  }
+push_end:
```
Comment (gkelly) [inline sw/device/tests/sha256_test.c:54]: Can you disentangle the flow here like below?

File: sw/device/boot_rom/bootstrap.c [hunk 4]
```diff
@@ -64,17 +64,54 @@ static int erase_flash(void) {
...
+
+  const uint8_t *data8 = data;
+  size_t data_left = len;
+  while (data_left > 0) {
+    size_t bytes_sent;
+    switch (dif_hmac_fifo_push(hmac, data8, data_left, &bytes_sent)) {
+      case kDifHmacFifoOk:
+      case kDifHmacFifoFull:
+        data8 += bytes_sent;
+        data_left -= bytes_sent;
```
Comment (gkelly) [inline sw/device/boot_rom/bootstrap.c:82]: This falls through. I don't like the loop-switch control flow here, but like the loop-switch-goto less from the other commit, can you make this more vanilla flow control?

File: sw/device/tests/sha256_test.c [hunk 5]
```diff
@@ -21,16 +27,54 @@ static const uint32_t kExpectedDigest[8] = {0xdc96c23d, 0xaf36e268, 0xcb68ff71,
...
+  const char *data = kData;
+  size_t data_left = kDataLen;
+  while (data_left > 0) {
+    size_t bytes_sent;
+    switch (dif_hmac_fifo_push(&hmac0, data, data_left, &bytes_sent)) {
+      case kDifHmacFifoOk:
+        goto push_end;
+      case kDifHmacFifoFull:
+        data += bytes_sent;
+        data_left -= bytes_sent;
```
Comment (gkelly) [inline sw/device/tests/sha256_test.c:48]: This falls through?

**Review by gkelly — CHANGES_REQUESTED** [review]

File: sw/device/boot_rom/bootstrap.c (hunk 4 above)
Comment (mcy) [reply sw/device/boot_rom/bootstrap.c:82]: Regardless of how I rewrite it, I'm surprised we don't have `-Wimplicit-fallthrough` turned on. We should absolutely turn it on.

File: sw/device/tests/sha256_test.c (hunk 2 above)
Comment (mcy) [reply sw/device/tests/sha256_test.c:77]: Done. I have *no idea* what this is here for, but I'm not going to delete this quite yet.

**Comment by lenary** [comment]
> Also, this seems to render sw/vendor/cryptoc completely unused. Should we delete it?

No, don't, as we may want to test SW implementations of crypto functionality against hardware ones, and it's useful to have this in the repo to be able to do so, especially if we know we can compile those files for Ibex.

**Review by moidx — APPROVED** [review]

**Pull request merged**